contracts/src/v0.8/PermissionedForwardProxy.sol

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./ConfirmedOwner.sol";
+import {getRevertMsg} from "./utils/utils.sol";
+
+/**
+ * @title PermissionedForwardProxy
+ * @notice This proxy is used to forward calls from sender to target. It maintains
+ * a permission list to check which sender is allowed to call which target
+ */
+contract PermissionedForwardProxy is ConfirmedOwner {
+  mapping(address => address) public forwardPermissionList;
+
+  constructor() ConfirmedOwner(msg.sender) {}
+
+  /**
+   * @notice Verifies if msg.sender has permission to forward to target address and then forwards the handler
+   * @param target address of the contract to forward the handler to
+   * @param handler bytes to be passed to target in call data
+   */
+  function forward(address target, bytes calldata handler) external {
+    require(forwardPermissionList[msg.sender] == target, "Forwarding permission not found");
```

Comment:
This allows the target to be 0 address which can be used for contract creation. I can add another check for 0 here but not sure if that's really required at the cost of more gas per tx. Thoughts?